base/special/gamma.jl

```diff
@@ -630,12 +604,90 @@ function eta(z::Union{Float64,Complex{Float64}})
         return -zeta(z) * expm1(0.6931471805599453094172321214581765*δz)
     end
 end
-eta(x::Integer) = eta(Float64(x))
-eta(x::Real)    = oftype(float(x),eta(Float64(x)))
 
-"""
-    eta(x)
 
-Dirichlet eta function ``\\eta(s) = \\sum^\\infty_{n=1}(-1)^{n-1}/n^{s}``.
-"""
-eta(z::Complex) = oftype(float(z),eta(Complex128(z)))
+# Converting types that we can convert, and not ones we can not
+# Float16, and Float32 and their Complex equivalents can be converted to Float64
+# and results converted back.
+# Otherwise, we need to make things use their own `float` converting methods
+# and in those cases, we do not convert back either as we assume
+# they also implement their own versions of the functions, with the correct return types.
+# This is the case for BitIntegers (which become `Float64` when `float`ed).
+# Otherwise, if they do not implement their version of the functions we
+# manually throw a `MethodError`.
+# This case occurs, when calling `float` on a type does not change its type,
+# as it is already a `float`, and would have hit own method, if one had existed.
+
+
+# If we really cared about single precision, we could make a faster
+# Float32 version by truncating the Stirling series at a smaller cutoff.
+# and if we really cared about half precision, we could make a faster
+# Float16 version, by using a precomputed table look-up.
+
+
+for T in (Float16, Float32, Float64)
+    @eval f64(x::Complex{$T}) = Complex128(x)
+    @eval f64(x::$T) = Float64(x)
+end
+
+
+for f in (:digamma, :trigamma, :zeta, :eta, :invdigamma)
+    @eval begin
+		function $f(z::Union{ComplexOrReal{Float16}, ComplexOrReal{Float32}})
+			oftype(z, $f(f64(z)))
+		end
+
+        function $f(z::Number)
+            x = float(z)
+            typeof(x) === typeof(z) && throw(MethodError($f, (z,)))
+            # There is nothing to fallback to, since this didn't change the argument types
+            $f(x)
+        end
+    end
+end
+
+
+for T1 in (Float16, Float32, Float64), T2 in (Float16, Float32, Float64)
+    (T1 == T2 == Float64) && continue # Avoid redefining base definition
+
+	@eval function zeta(s::ComplexOrReal{$T1}, z::ComplexOrReal{$T2})
+		ζ = zeta(f64(s), f64(z))
+
+		if typeof(s) <: Complex || typeof(z) <: Complex
+			convert(Complex{$T2}, ζ)
```

Comment:
It used to be, but after thinking a bit more about `zeta`, 
I started to think that it should maybe be driven only by the type `$T2`.
But I now think I was going down the wrong line.

 - I wanted `zeta(::BigInteger, ::Float64) :: Float64`   (which is what depending only on `$T2` gives me)
 - `promote_type($T1,$T2}`, gives me `zeta(::BigInteger, ::Float64) :: BigFloat` -- which is a lie as that precision is not there.
 - I now think that it should just `MethodError`, which I can get by deleting a definition (or two)